src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs

```diff
@@ -306,47 +272,45 @@ private bool TryBindUnconvertedBinaryOperatorToDefaultInterpolatedStringHandler(
+        private BoundBinaryOperator UpdateBinaryOperatorWithInterpolatedContents(BoundBinaryOperator originalOperator, ImmutableArray<ImmutableArray<BoundExpression>> appendCalls, InterpolatedStringHandlerData data, SyntaxNode rootSyntax, BindingDiagnosticBag diagnostics)
         {
-            Debug.Assert(appendCalls.Length == stack.Count + 1);
             var @string = GetSpecialType(SpecialType.System_String, diagnostics, rootSyntax);
 
-            var bottomOperator = stack.Pop();
-            var result = createBinaryOperator(bottomOperator, createInterpolation(bottomOperator.Left, appendCalls[0]), rightIndex: 1);
+            Func<(ImmutableArray<ImmutableArray<BoundExpression>>, TypeSymbol), BoundUnconvertedInterpolatedString, int, BoundExpression> interpolationFactory =
+                createInterpolation;
+            Func<(ImmutableArray<ImmutableArray<BoundExpression>>, TypeSymbol), BoundBinaryOperator, BoundExpression, BoundExpression, BoundExpression> binaryOperatorFactory =
+                createBinaryOperator;
 
-            for (int i = 2; i < appendCalls.Length; i++)
-            {
-                result = createBinaryOperator(stack.Pop(), result, rightIndex: i);
-            }
+            var rewritten = (BoundBinaryOperator)originalOperator.RewriteInterpolatedStringAddition((appendCalls, @string), interpolationFactory, binaryOperatorFactory);
 
-            return result.Update(BoundBinaryOperator.UncommonData.InterpolatedStringHandlerAddition(data));
+            return rewritten.Update(BoundBinaryOperator.UncommonData.InterpolatedStringHandlerAddition(data));
 
-            BoundBinaryOperator createBinaryOperator(BoundBinaryOperator original, BoundExpression left, int rightIndex)
+            static BoundBinaryOperator createBinaryOperator((ImmutableArray<ImmutableArray<BoundExpression>> _, TypeSymbol @string) arg, BoundBinaryOperator original, BoundExpression left, BoundExpression right)
                 => new BoundBinaryOperator(
                     original.Syntax,
                     BinaryOperatorKind.StringConcatenation,
                     left,
-                    createInterpolation(original.Right, appendCalls[rightIndex]),
+                    right,
                     original.ConstantValue,
                     methodOpt: null,
                     constrainedToTypeOpt: null,
                     LookupResultKind.Viable,
                     originalUserDefinedOperatorsOpt: default,
-                    @string,
+                    arg.@string,
                     original.HasErrors);
 
-            static BoundInterpolatedString createInterpolation(BoundExpression expression, ImmutableArray<BoundExpression> parts)
+            static BoundInterpolatedString createInterpolation((ImmutableArray<ImmutableArray<BoundExpression>> AppendCalls, TypeSymbol _) arg, BoundExpression expression, int i)
```

Comment:
The type of the `expression` parameter could be made `BoundUnconvertedInterpolatedString` (narrower).
nit: for readability, consider moving this local function above `createBinaryOperator` (ie. swap the two local functions) to match usage order in the call to `RewriteInterpolatedStringAddition`
nit: consider moving `arg` to be the last parameter. That's more typical. Something like `Func<TInterpolatedStringType, int, TArg, TResult> interpolatedStringFactory` and `Func<BoundBinaryOperator, TResult, TResult, TArg, TResult> binaryOperatorFactory` in `RewriteInterpolatedStringAddition`